collectors/delta_counter.go

```diff
@@ -0,0 +1,150 @@
+// Copyright 2020 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package collectors
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"google.golang.org/api/monitoring/v3"
+)
+
+type CollectedMetric struct {
+	metric          *ConstMetric
+	lastCollectedAt time.Time
+}
+
+// DeltaCounterStore defines a set of functions which must be implemented in order to be used as a DeltaCounterStore
+// which accumulates DELTA Counter metrics over time
+type DeltaCounterStore interface {
+
+	// Increment will use the incoming metricDescriptor and currentValue to either create a new entry or add the incoming
+	// value to an existing entry in the underlying store
+	Increment(metricDescriptor *monitoring.MetricDescriptor, currentValue *ConstMetric)
+
+	// ListMetrics will return all known entries in the store for a metricDescriptorName
+	ListMetrics(metricDescriptorName string) map[string][]*CollectedMetric
+}
+
+type metricEntry = map[uint64]*CollectedMetric
+
+type inMemoryDeltaCounterStore struct {
+	store      map[string]metricEntry
+	ttl        time.Duration
+	storeMutex *sync.RWMutex
+	logger     log.Logger
+}
+
+// NewInMemoryDeltaCounterStore returns an implementation of DeltaCounterStore which is persisted in-memory
+func NewInMemoryDeltaCounterStore(logger log.Logger, ttl time.Duration) DeltaCounterStore {
+	return inMemoryDeltaCounterStore{
+		store:      map[string]metricEntry{},
+		storeMutex: &sync.RWMutex{},
+		logger:     logger,
+		ttl:        ttl,
+	}
+}
+
+func (s inMemoryDeltaCounterStore) Increment(metricDescriptor *monitoring.MetricDescriptor, currentValue *ConstMetric) {
+	if currentValue == nil {
+		return
+	}
+
+	var entry metricEntry
+	s.storeMutex.Lock()
+	if _, exists := s.store[metricDescriptor.Name]; !exists {
+		s.store[metricDescriptor.Name] = metricEntry{}
+	}
+	entry = s.store[metricDescriptor.Name]
+	s.storeMutex.Unlock()
+
+	key := toCounterKey(currentValue)
+	existing := entry[key]
+
+	if existing == nil {
+		level.Debug(s.logger).Log("msg", "Tracking new counter", "fqName", currentValue.fqName, "key", key, "current_value", currentValue.value, "incoming_time", currentValue.reportTime)
+		entry[key] = &CollectedMetric{currentValue, time.Now()}
```

Comment:
Maybe out of context question, but isn't it possible for two or more routines to get to this point with the same `entry` reference?

Reply:
So, this is actually cheating a bit with knowing some of the external context for how things are run. The current parallelism is based on MetricDescriptor so it's _technically_ safe.